dotcom-rendering/src/client/userFeatures/user-features.ts

```diff
@@ -0,0 +1,170 @@
+// This file was migrated from:
+// https://github.com/guardian/commercial/blob/1a429d6be05657f20df4ca909df7d01a5c3d7402/src/lib/user-features.ts
+import { getCookie, isObject, removeCookie, setCookie } from '@guardian/libs';
+import {
+	getAuthStatus,
+	getOptionsHeadersWithOkta,
+	isUserLoggedInOktaRefactor,
+} from '../../lib/identity';
+import {
+	adFreeDataIsPresent,
+	cookieIsExpiredOrMissing,
+	fetchJson,
+	getAdFreeCookie,
+	setAdFreeCookie,
+	timeInDaysFromNow,
+} from './user-features-lib';
+import type { UserFeaturesResponse } from './user-features-lib';
+
+const USER_FEATURES_EXPIRY_COOKIE = 'gu_user_features_expiry';
+const PAYING_MEMBER_COOKIE = 'gu_paying_member';
+const ACTION_REQUIRED_FOR_COOKIE = 'gu_action_required_for';
+const DIGITAL_SUBSCRIBER_COOKIE = 'gu_digital_subscriber';
+const HIDE_SUPPORT_MESSAGING_COOKIE = 'gu_hide_support_messaging';
+const AD_FREE_USER_COOKIE = 'GU_AF1';
+
+const RECURRING_CONTRIBUTOR_COOKIE = 'gu_recurring_contributor';
+const ONE_OFF_CONTRIBUTION_DATE_COOKIE = 'gu_one_off_contribution_date';
+
+const forcedAdFreeMode = !!/[#&]noadsaf(&.*)?$/.exec(window.location.hash);
+
+const userHasData = () => {
+	const cookie =
+		getAdFreeCookie() ??
+		getCookie({ name: ACTION_REQUIRED_FOR_COOKIE }) ??
+		getCookie({ name: USER_FEATURES_EXPIRY_COOKIE }) ??
+		getCookie({ name: PAYING_MEMBER_COOKIE }) ??
+		getCookie({ name: RECURRING_CONTRIBUTOR_COOKIE }) ??
+		getCookie({ name: ONE_OFF_CONTRIBUTION_DATE_COOKIE }) ??
+		getCookie({ name: DIGITAL_SUBSCRIBER_COOKIE }) ??
+		getCookie({ name: HIDE_SUPPORT_MESSAGING_COOKIE });
+	return !!cookie;
+};
+
+const validateResponse = (
+	response: unknown,
+): response is UserFeaturesResponse => {
+	if (!isObject(response)) return false;
+
+	if (
+		typeof response.showSupportMessaging === 'boolean' &&
+		isObject(response.contentAccess) &&
+		typeof response.contentAccess.paidMember === 'boolean' &&
+		typeof response.contentAccess.recurringContributor === 'boolean' &&
+		typeof response.contentAccess.digitalPack === 'boolean'
+	) {
+		return true;
+	}
+
+	return true;
```

Comment:
That’s an odd fallback, especially since we’re making all these checks above.

Reply:
Good spot, have addressed and updated related tests.